File: leetcode/1723.完成所有工作的最短时间.py
subsets = [[] for _ in range(1 << 12)]
for i in range(1 << 12):
    s = i
    while s:
        subsets[i].append(s)
        s = (s - 1) & i
class Solution:
    def minimumTimeRequired(self, jobs: List[int], k: int) -> int:
        m = 1 << len(jobs)
        pre = [0] * m
        for i in range(1, m):
            pre[i] = pre[i^i&-i] + jobs[(i&-i).bit_length()-1]
        f = [[0] * m for _ in range(k)]
        for i in range(m): f[0][i] = pre[i]
        for i in range(1, k):
            for j in range(1, m):
                t = inf
                for s in subsets[j]:
                    v = f[i-1][j^s]
                    if pre[s] > v: v = pre[s]
                    if v < t: t = v
                f[i][j] = t
        return f[k-1][m-1]

File: leetcode/1723.完成所有工作的最短时间.cpp
class Solution {
public:
    const int MX = 1e8;
    int minimumTimeRequired(vector<int>& jobs, int k) {
        int n = jobs.size();
        int m = 1 << n;
        vector<int> pre(m, 0);
        for(int i = 1; i < m; i++){
            pre[i] = pre[i^i&-i] + jobs[__builtin_ctz(i)];
        }
        vector<vector<int>> f(k, vector<int>(m, 0));
        for(int i = 0; i < m; i++) f[0][i] = pre[i];
        for(int i = 1; i < k; i++){
            for(int j = 1; j < m; j++){
                int v = MX;
                for(int s = j; s; s = (s - 1) & j){
                    v = min(v, max(pre[s], f[i-1][j^s]));
                }
                f[i][j] = v;
            }
        }
        return f[k-1][m-1];
    }
};